librabbitmq/amqp_socket.c

```diff
@@ -1262,14 +1316,24 @@ static amqp_rpc_reply_t amqp_login_inner(amqp_connection_state_t state,
   }
   client_heartbeat = (uint16_t)heartbeat;
 
-  res = amqp_send_header(state);
+  res = amqp_time_from_now(&deadline, timeout);
   if (AMQP_STATUS_OK != res) {
     goto error_res;
   }
 
-  res = amqp_simple_wait_method(state, 0, AMQP_CONNECTION_START_METHOD,
-                                &method);
-  if (res != AMQP_STATUS_OK) {
+  res = amqp_send_header_noblock(state, deadline);
+  if (AMQP_STATUS_OK != res) {
+    goto error_res;
+  }
+
+  res = amqp_time_tv_until(deadline, &tv, &tvp);
```

Comment:
Change the functions called by this function so that they take a deadline and not a timeout, then it won't be necessary to call this repeatedly during the function.